harvester/create.go

```diff
@@ -86,51 +86,65 @@ func (d *Driver) PreCreateCheck() error {
 }
 
 func (d *Driver) Create() error {
-	c, err := d.getClient()
+	// create keypair
+	if err := d.createKeyPair(); err != nil {
+		return err
+	}
+	// create vm
+	cloudInitSource, cloudConfigSecret, err := d.buildCloudInit()
 	if err != nil {
 		return err
 	}
-
-	if err = d.createKeyPair(); err != nil {
+	imageNamespace, imageName, err := NamespacedNamePartsByDefault(d.ImageName, d.VMNamespace)
+	if err != nil {
 		return err
 	}
-
-	userData, networkData := d.createCloudInit()
-
 	dataVolumeOption := &builder.DataVolumeOption{
+		ImageID:          fmt.Sprintf("%s/%s", imageNamespace, imageName),
 		VolumeMode:       corev1.PersistentVolumeBlock,
 		AccessMode:       corev1.ReadWriteMany,
-		StorageClassName: pointer.StringPtr("longhorn-" + d.ImageName),
-		ImageID:          fmt.Sprintf("%s/%s", d.Namespace, d.ImageName),
+		StorageClassName: pointer.StringPtr(builder.BuildImageStorageClassName("", imageName)),
 	}
-
-	// create vm
 	vmBuilder := builder.NewVMBuilder("docker-machine-driver-harvester").
-		Namespace(d.Namespace).Name(d.MachineName).
-		CPU(d.CPU).Memory(d.MemorySize).
-		Image(d.DiskSize, d.DiskBus, dataVolumeOption).
-		EvictionStrategy(true).
-		DefaultPodAntiAffinity().
-		CloudInit(userData, networkData)
+		Namespace(d.VMNamespace).Name(d.MachineName).CPU(d.CPU).Memory(d.MemorySize).
+		DataVolumeDisk(rootDiskName, builder.DiskBusVirtio, false, 1, d.DiskSize, "", dataVolumeOption).
+		CloudInitDisk(builder.CloudInitDiskName, builder.DiskBusVirtio, false, 0, *cloudInitSource).
+		EvictionStrategy(true).DefaultPodAntiAffinity().Run(false)
 
 	if d.KeyPairName != "" {
 		vmBuilder = vmBuilder.SSHKey(d.KeyPairName)
 	}
-
-	if d.NetworkType != networkTypePod {
-		vmBuilder = vmBuilder.Bridge(d.NetworkName, d.NetworkModel)
-	} else {
-		vmBuilder = vmBuilder.ManagementNetwork(true)
+	interfaceType := builder.NetworkInterfaceTypeBridge
+	networkName := d.NetworkName
+	if d.NetworkType == networkTypePod {
+		networkName = ""
 	}
-
-	if _, err = c.VirtualMachines.Create(vmBuilder.Run()); err != nil {
+	vm, err := vmBuilder.NetworkInterface(interfaceName, d.NetworkModel, "", interfaceType, networkName).VM()
+	if err != nil {
 		return err
 	}
-
-	if err = d.waitForState(state.Running); err != nil {
+	vm.Kind = kubevirtv1.VirtualMachineGroupVersionKind.Kind
+	vm.APIVersion = kubevirtv1.GroupVersion.String()
+	createdVM, err := d.createVM(vm)
+	if err != nil {
 		return err
 	}
-	if err = d.waitForIP(); err != nil {
+	// create secret
+	if cloudConfigSecret != nil {
+		cloudConfigSecret.OwnerReferences = []metav1.OwnerReference{
+			{
+				APIVersion: vm.APIVersion,
+				Kind:       vm.Kind,
+				Name:       vm.Name,
+				UID:        createdVM.UID,
+			},
+		}
+		if _, err = d.createSecret(cloudConfigSecret); err != nil {
+			return err
+		}
+	}
```

Comment:
does it has an owner reference, make sure when the VM is deleted the related secret is also removed.

Reply:
yes